node/lib/util/synthetic_gc_runner.js

```diff
@@ -0,0 +1,473 @@
+
+        const fetcher = new SubmoduleFetcher(repo, refHeadCommit);
+
+        let subUrl = yield fetcher.getSubmoduleUrl(subName);
+        let subPath = subUrl;
+
+        if (subUrl.startsWith(".")) {
+            // relative url specified
+            subUrl = detail.rebuildUrl(subUrl);
+            subPath = yield SyntheticBranchUtil.urlToLocalPath(repo, subUrl);
+        }
+
+        if (!fs.existsSync(subPath)) {
+
+            if (this.d_verbose) {
+                console.log("Submodule at following path does not exist: " +
+                    subPath);
+            }
+
+            const metaUrl = yield GitUtil.getOriginUrl(repo);
+            if (metaUrl === null) {
+                console.error("Cannot determine origin url. Something is " +
+                    "wrong, exiting.");
+                process.exit(-1);
+            }
+
+            subUrl = yield fetcher.getSubmoduleUrl(subName);
+            subUrl = SubmoduleConfigUtil.resolveSubmoduleUrl(metaUrl,
+                                                             subUrl);
+            if (this.d_verbose) {
+                console.log("Going to clone bare repo from: " + subUrl);
+            }
+            yield GitUtil.cloneBareRepo(subUrl, subPath);
+        }
+
+        const subRepo = yield NodeGit.Repository.open(subPath);
+        try {
+            if (!(subUrl in this.d_fetchedUrl)) {
+                this.d_fetchedUrl[subUrl] = 1;
+                yield GitUtil.fetch(subRepo, "origin");
+            }
+        } catch (exception) {
+            // eat the exception here, most likely submodule is corrupted.
+            // 'populatePerCommit' has more infromative error for this
+            console.log("Error fetching submodule: " + subName +
+                " with error: " + exception);
+        }
+
+        return subRepo;
+});
+
+/**
+ * Execute batch removal of synthetic references contained by
+ * 'd_syntheticRefsBatchForRemoval' for the specified 'repo'. Lastly, clean up
+ * 'd_syntheticRefsBatchForRemoval'.
+ *
+ * @param {NodeGit.Repository}   repo
+ */
+SyntheticGcRunner.prototype.commitSyntheticRefRemoval = co.wrap(
+    function*(repo) {
+
+   yield GitUtil.removeRemoteRef(repo,
+                                 this.d_syntheticRefsBatchForRemoval);
+   this.d_syntheticRefsBatchForRemoval = [];
+
+});
+
+/**
+ * Remove synthetic ref corresponding to specified `commit` in the specified
+ * `repo`.
+ * If running outside of git server, synthetic ref will be marked for batch
+ * removal, you will need to call 'SyntheticGcRunner.commitSyntheticRefRemoval'
+ * to actually remove them.
+ *
+ * @param {NodeGit.Repository}   repo
+ * @param {NodeGit.Commit}       commit
+ */
+SyntheticGcRunner.prototype.removeSyntheticRef = function(repo, commit) {
+
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(commit, NodeGit.Commit);
+
+    const synRefPath = SyntheticBranchUtil.getSyntheticBranchForCommit(commit);
+
+    if (this.d_simulation) {
```

Comment:
This seems a little odd to me; I'd rather not call this method when simulating than to have it check for simulation.